host/settings/pin/pin.go

```diff
@@ -0,0 +1,262 @@
+		// host's settings are updated based on the current exchange rate.
+		Threshold float64 `json:"threshold"`
+
+		// Storage, Ingress, and Egress are the pinned prices in the
+		// external currency.
+		Storage float64 `json:"storage"`
+		Ingress float64 `json:"ingress"`
+		Egress  float64 `json:"egress"`
+
+		// MaxCollateral is the maximum collateral that the host will
+		// accept in the external currency.
+		MaxCollateral float64 `json:"maxCollateral"`
+	}
+
+	// A SettingsManager updates and retrieves the host's settings.
+	SettingsManager interface {
+		Settings() settings.Settings
+		UpdateSettings(settings.Settings) error
+	}
+
+	// A Store stores and retrieves pinned settings.
+	Store interface {
+		PinnedSettings() (PinnedSettings, error)
+		UpdatePinnedSettings(PinnedSettings) error
+	}
+
+	// An ExchangeRateRetriever retrieves the current exchange rate from
+	// an external source.
+	ExchangeRateRetriever interface {
+		SiacoinExchangeRate(ctx context.Context, currency string) (float64, error)
+	}
+
+	// A Manager manages the host's pinned settings and updates the host's
+	// settings based on the current exchange rate.
+	Manager struct {
+		log      *zap.Logger
+		store    Store
+		explorer ExchangeRateRetriever
+		sm       SettingsManager
+
+		frequency  time.Duration
+		rateWindow time.Duration
+
+		mu       sync.Mutex
+		rates    []decimal.Decimal
+		lastRate decimal.Decimal
+		pinned   PinnedSettings // in-memory cache of pinned settings
+	}
+)
+
+func isOverThreshold(a, b, percentage decimal.Decimal) bool {
+	threshold := a.Mul(percentage)
+	diff := a.Sub(b).Abs()
+	return diff.GreaterThanOrEqual(threshold)
+}
+
+func convertToCurrency(target decimal.Decimal, rate decimal.Decimal) types.Currency {
+	hastings := target.Div(rate).Mul(decimal.New(1, 24)).Round(0).String()
+	c, err := types.ParseCurrency(hastings)
+	if err != nil {
+		panic(err)
+	}
+	return c
+}
+
+func averageRate(rates []decimal.Decimal) decimal.Decimal {
+	var sum decimal.Decimal
+	for _, r := range rates {
+		sum = sum.Add(r)
+	}
+	return sum.Div(decimal.NewFromInt(int64(len(rates))))
+}
+
+func (m *Manager) updatePrices(ctx context.Context, force bool) error {
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	m.mu.Lock()
+	currency := m.pinned.Currency
+	m.mu.Unlock()
+
+	if currency == "" {
+		return nil
+	}
+
+	rate, err := m.explorer.SiacoinExchangeRate(ctx, currency)
+	if err != nil {
+		return fmt.Errorf("failed to get exchange rate: %w", err)
+	} else if rate <= 0 {
+		return fmt.Errorf("exchange rate must be positive")
+	}
+	current := decimal.NewFromFloat(rate)
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	maxRates := int(m.rateWindow / m.frequency)
+	m.rates = append(m.rates, current)
+	if len(m.rates) >= maxRates {
+		m.rates = m.rates[1:]
+	}
+
+	// skip updating prices if the pinned settings are zero
+	if m.pinned.Storage <= 0 && m.pinned.Ingress <= 0 && m.pinned.Egress <= 0 && m.pinned.MaxCollateral <= 0 {
```

Comment:
Might be a bit more robust to be explicit about it and use a boolean to enable/disable pinning.
I feel like it might make sense to be able to pin e.g. storage to $0 when pinning is enabled versus considering pinning for storage to be disabled when it is set to 0 and having to update the regular price settings.

Reply:
There's no point to pin to 0. 0 is 0 no matter what currency you're in